File: badminton_pipeline/calibration/homography.py
"""Versioned four-corner court calibration using only the standard library."""

from __future__ import annotations

import argparse
import json
import math
import os
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Sequence


Point = tuple[float, float]
COURT_POINTS: tuple[Point, ...] = ((0.0, 0.0), (6.1, 0.0), (6.1, 13.4), (0.0, 13.4))


def solve_linear_system(rows: Sequence[Sequence[float]]) -> list[float]:
    size = len(rows)
    matrix = [list(row) for row in rows]
    for column in range(size):
        pivot = max(range(column, size), key=lambda row: abs(matrix[row][column]))
        if abs(matrix[pivot][column]) < 1e-10:
            raise ValueError("Degenerate corner geometry; homography is not solvable")
        matrix[column], matrix[pivot] = matrix[pivot], matrix[column]
        divisor = matrix[column][column]
        matrix[column] = [value / divisor for value in matrix[column]]
        for row in range(size):
            if row == column:
                continue
            factor = matrix[row][column]
            for index in range(column, size + 1):
                matrix[row][index] -= factor * matrix[column][index]
    return [matrix[row][size] for row in range(size)]


def compute_homography(source: Sequence[Point]) -> list[list[float]]:
    if len(source) != 4:
        raise ValueError("Exactly four court corners are required")
    rows: list[list[float]] = []
    for (x, y), (u, v) in zip(source, COURT_POINTS, strict=True):
        rows.append([x, y, 1.0, 0.0, 0.0, 0.0, -u * x, -u * y, u])
        rows.append([0.0, 0.0, 0.0, x, y, 1.0, -v * x, -v * y, v])
    values = solve_linear_system(rows)
    return [values[0:3], values[3:6], [values[6], values[7], 1.0]]


def project_point(matrix: Sequence[Sequence[float]], point: Point) -> Point:
    x, y = point
    denominator = matrix[2][0] * x + matrix[2][1] * y + matrix[2][2]
    if abs(denominator) < 1e-12:
        raise ValueError("Point projects to infinity")
    return (
        (matrix[0][0] * x + matrix[0][1] * y + matrix[0][2]) / denominator,
        (matrix[1][0] * x + matrix[1][1] * y + matrix[1][2]) / denominator,
    )


def validate_corners(points: Sequence[Point], width: int, height: int) -> tuple[bool, float, str]:
    if len(points) != 4:
        return False, 0.0, "Exactly four court corners are required"
    if width <= 0 or height <= 0:
        return False, 0.0, "Image dimensions must be positive"
    if any(not math.isfinite(x) or not math.isfinite(y) or x < 0 or y < 0 or x > width or y > height for x, y in points):
        return False, 0.0, "All corners must be inside the image"

    signed_area = sum(
        points[index][0] * points[(index + 1) % 4][1]
        - points[(index + 1) % 4][0] * points[index][1]
        for index in range(4)
    ) / 2.0
    area_ratio = abs(signed_area) / (width * height)
    crosses = []
    for index in range(4):
        point, nxt, after = points[index], points[(index + 1) % 4], points[(index + 2) % 4]
        crosses.append((nxt[0] - point[0]) * (after[1] - nxt[1]) - (nxt[1] - point[1]) * (after[0] - nxt[0]))
    if not (all(value > 0 for value in crosses) or all(value < 0 for value in crosses)):
        return False, area_ratio, "Court corners cross or form a concave quadrilateral"
    if area_ratio < 0.015:
        return False, area_ratio, "Court quadrilateral is too small"
    if (points[0][1] + points[1][1]) / 2 >= (points[2][1] + points[3][1]) / 2:
        return False, area_ratio, "Upper corners must be above lower corners"
    return True, area_ratio, "accepted"


def write_json(path: Path, value: object) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    temporary = path.with_suffix(path.suffix + ".tmp")
    with temporary.open("w", encoding="utf-8", newline="\n") as stream:
        json.dump(value, stream, ensure_ascii=False, indent=2)
        stream.write("\n")
    os.replace(temporary, path)


def parse_corners(value: str) -> list[Point]:
    points: list[Point] = []
    for item in value.split(";"):
        x, y = item.split(",", 1)
        points.append((float(x), float(y)))
    return points


def main() -> int:
    parser = argparse.ArgumentParser(description="Save a versioned manual court calibration.")
    parser.add_argument("--video-id", required=True)
    parser.add_argument("--corners", required=True, help="ULx,ULy;URx,URy;LRx,LRy;LLx,LLy")
    parser.add_argument("--width", required=True, type=int)
    parser.add_argument("--height", required=True, type=int)
    parser.add_argument("--frame-time-ms", default=0, type=int)
    parser.add_argument("--data-dir", default=Path("data"), type=Path)
    args = parser.parse_args()

    corners = parse_corners(args.corners)
    valid, area_ratio, message = validate_corners(corners, args.width, args.height)
    if not valid:
        parser.error(message)
    homography = compute_homography(corners)
    output_dir = args.data_dir.resolve() / "calibrations" / args.video_id
    existing = sorted(output_dir.glob("v*.json")) if output_dir.exists() else []
    version = max((int(path.stem.removeprefix("v")) for path in existing), default=0) + 1
    record = {
        "calibration_id": f"cal_{uuid.uuid4()}",
        "video_id": args.video_id,
        "version": version,
        "frame_time_ms": max(0, args.frame_time_ms),
        "image_width": args.width,
        "image_height": args.height,
        "corners": [{"x": x, "y": y} for x, y in corners],
        "homography": homography,
        "court_orientation": "upper_is_far_side",
        "quality_status": "accepted",
        "quality": {"area_ratio": area_ratio},
        "source": "manual",
        "created_at": datetime.now(timezone.utc).isoformat(timespec="seconds"),
    }
    output = output_dir / f"v{version}.json"
    write_json(output, record)
    print(json.dumps({"status": "saved", "version": version, "path": str(output), "homography": homography}, ensure_ascii=False, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: demo/lib/homography.ts
export type CalibrationPoint = { x: number; y: number };

export type CalibrationValidation = {
  valid: boolean;
  areaRatio: number;
  message: string;
};

const COURT_POINTS: CalibrationPoint[] = [
  { x: 0, y: 0 },
  { x: 6.1, y: 0 },
  { x: 6.1, y: 13.4 },
  { x: 0, y: 13.4 },
];

function solveLinearSystem(rows: number[][]): number[] {
  const size = rows.length;
  const matrix = rows.map((row) => [...row]);
  for (let column = 0; column < size; column += 1) {
    let pivot = column;
    for (let row = column + 1; row < size; row += 1) {
      if (Math.abs(matrix[row][column]) > Math.abs(matrix[pivot][column])) pivot = row;
    }
    if (Math.abs(matrix[pivot][column]) < 1e-10) throw new Error("角点几何退化，无法计算 Homography");
    [matrix[column], matrix[pivot]] = [matrix[pivot], matrix[column]];
    const divisor = matrix[column][column];
    for (let index = column; index <= size; index += 1) matrix[column][index] /= divisor;
    for (let row = 0; row < size; row += 1) {
      if (row === column) continue;
      const factor = matrix[row][column];
      for (let index = column; index <= size; index += 1) matrix[row][index] -= factor * matrix[column][index];
    }
  }
  return matrix.map((row) => row[size]);
}

export function computeHomography(source: CalibrationPoint[]): number[][] {
  if (source.length !== 4) throw new Error("需要四个球场角点");
  const rows: number[][] = [];
  source.forEach(({ x, y }, index) => {
    const { x: u, y: v } = COURT_POINTS[index];
    rows.push([x, y, 1, 0, 0, 0, -u * x, -u * y, u]);
    rows.push([0, 0, 0, x, y, 1, -v * x, -v * y, v]);
  });
  const values = solveLinearSystem(rows);
  return [values.slice(0, 3), values.slice(3, 6), [...values.slice(6, 8), 1]];
}

export function projectPoint(matrix: number[][], point: CalibrationPoint): CalibrationPoint {
  const denominator = matrix[2][0] * point.x + matrix[2][1] * point.y + matrix[2][2];
  return {
    x: (matrix[0][0] * point.x + matrix[0][1] * point.y + matrix[0][2]) / denominator,
    y: (matrix[1][0] * point.x + matrix[1][1] * point.y + matrix[1][2]) / denominator,
  };
}

export function validateCalibration(points: CalibrationPoint[], width: number, height: number): CalibrationValidation {
  if (points.length !== 4) return { valid: false, areaRatio: 0, message: `还需点击 ${4 - points.length} 个角点` };
  if (width <= 0 || height <= 0) return { valid: false, areaRatio: 0, message: "视频分辨率不可用" };
  if (points.some((point) => !Number.isFinite(point.x) || !Number.isFinite(point.y) || point.x < 0 || point.y < 0 || point.x > width || point.y > height)) {
    return { valid: false, areaRatio: 0, message: "角点必须位于视频画面内" };
  }

  const signedArea = points.reduce((sum, point, index) => {
    const next = points[(index + 1) % points.length];
    return sum + point.x * next.y - next.x * point.y;
  }, 0) / 2;
  const areaRatio = Math.abs(signedArea) / (width * height);
  const crosses = points.map((point, index) => {
    const next = points[(index + 1) % 4];
    const after = points[(index + 2) % 4];
    return (next.x - point.x) * (after.y - next.y) - (next.y - point.y) * (after.x - next.x);
  });
  const convex = crosses.every((value) => value > 0) || crosses.every((value) => value < 0);
  if (!convex) return { valid: false, areaRatio, message: "四个角点发生交叉，请按指定顺序重新点击" };
  if (areaRatio < 0.015) return { valid: false, areaRatio, message: "标定区域过小，请确认选择了球场外侧四角" };
  const upperAverage = (points[0].y + points[1].y) / 2;
  const lowerAverage = (points[2].y + points[3].y) / 2;
  if (upperAverage >= lowerAverage) return { valid: false, areaRatio, message: "上方角点必须位于下方角点之上" };
  return { valid: true, areaRatio, message: "几何检查通过，可以保存" };
}
